walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,94 @@
+import java.util.*;
 
 public final class FindMeetingQuery {
+
+  private static boolean attendanceRequired(Event event, MeetingRequest request) {
+     Set<String> s = event.getAttendees();
+     Collection<String> c = request.getAttendees();
+     for (String person : c) {
+       if (s.contains(person)){ return true; }
+     }
+     return false;
+  }
+
+  /**
+   * Returns a list of times that are compatible with a given list of events
+   * @param {Collection<Event>} events A list of events where each event contains
+   * information about when a specific group of people are meeting.
+   * @param {MeetingRequest} request An object describing the meeting being requested.
+   * @return {Collection<TimeRange>} A list of meeting times where all required 
+   * attendees are available
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Make two sorted lists, based on meeting start and end time.
+    List<TimeRange> busyTimesStart = new ArrayList<TimeRange>();
+    List<TimeRange> busyTimesEnd = new ArrayList<TimeRange>();
+    for (Event event : events) {
+      if (attendanceRequired(event, request)){
+        busyTimesStart.add(event.getWhen());
+        busyTimesEnd.add(event.getWhen());
+      }
+    }
+    Collections.sort(busyTimesStart, TimeRange.ORDER_BY_START);
+    Collections.sort(busyTimesEnd, TimeRange.ORDER_BY_END);
+
+    // Store a list of times when no attendees are busy
+    List<TimeRange> allAvailableTimes = new ArrayList<TimeRange>();
+    
+    // Return empty list if requested event is longer than the whole day
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) { return allAvailableTimes; }
+
+    // If no required attendees are busy, return the whole day for availability
+    int totalEvents = busyTimesStart.size();
+    if (totalEvents == 0) {
+      allAvailableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+      return allAvailableTimes;
+    }
+
+    // Check for free time before first scheduled event.
+    int first = 0;
+    if (busyTimesStart.get(first).start() > TimeRange.START_OF_DAY) {
+      allAvailableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, busyTimesStart.get(first).start(), false));
+    }
+
+    int curStartIndex = 0, curEndIndex = 1, lastEventIndex = totalEvents - 1;
+
+    while (curStartIndex < totalEvents && curEndIndex < totalEvents) {
+      while (curEndIndex < lastEventIndex && busyTimesStart.get(curStartIndex).overlaps(busyTimesEnd.get(curEndIndex))) {
+        ++curEndIndex;
+      }
+
+      if (curEndIndex == lastEventIndex) { 
+        if (busyTimesStart.get(curStartIndex).end() < busyTimesEnd.get(curEndIndex).start()) {
+          allAvailableTimes.add(TimeRange.fromStartEnd(busyTimesStart.get(curStartIndex).end(), busyTimesEnd.get(curEndIndex).start(), false));
+        }
```

Comment:
```
busyTimesStart: [[1, 5], [2, 4], [3, 5], [4, 6], [11, 12]],
busyTimesEnd: [[2, 4], [1, 5], [3, 5], [4, 6], [11, 12]],
```
Not 100% on this, but won't this result in us executing
```
allAvailableTimes.add(TimeRange.fromStartEnd(5:00, 11:00))
```
?

I don't think any of the tests are breaking since they only have max 2 events and this scenario never occurs.